### PR TITLE
feat: implement jws for COSE refactor

### PR DESCRIPTION
**Pull request by JeyJeyGao**
Signed-off-by: Junjie Gao <junjiegao@microsoft.com>

### CODE REVIEW

File: signature/errors.go [hunk 1]
```diff
@@ -46,3 +46,40 @@ func (e *MalformedArgumentError) Error() string {
...
+	Alg string
+}
+
+func (e SignatureAlgoNotSupportedError) Error() string {
+	return fmt.Sprintf("signature algorithm %q is not supported", e.Alg)
+}
+
+// SignatureIntegrityError is used when the Signature associated is no longer valid.
+type SignatureIntegrityError struct {
+	Err error
```
Comment (binbin-li) [inline signature/errors.go:73]: can you implement `Unwrap()` as well?

File: signature/jws/jws.go [hunk 2]
```diff
@@ -25,24 +35,374 @@ func NewEnvelope() signature.Envelope {
...
+	}
+	return finalSig, certs, err
+}
+
+func validateCertificateChain(certChain []*x509.Certificate, expectedAlg signature.Algorithm, f func(string) error) error {
+	if len(certChain) == 0 {
+		return f("certificate-chain not present or is empty")
+	}
+
+	err := validateCertChain(certChain)
```
Comment (patrickzheng200) [inline signature/jws/jws.go:204]: Do we need to implement the cert chain check details again? see https://github.com/notaryproject/notation-core-go/blob/0553376c4e5d86681c177cbb5bfec37024dd69f6/signature/internal/base/envelope.go#L172

File: signature/jws/jws.go [hunk 3]
```diff
@@ -25,24 +35,374 @@ func NewEnvelope() signature.Envelope {
...
+	resSignAlgo := keySpec.SignatureAlgorithm()
+	if resSignAlgo != expectedAlg {
+		return f("mismatch between signature algorithm derived from signing certificate and signing algorithm specified")
+	}
+
+	return nil
+}
+
+// validateCertChain takes an ordered certificate chain and validates issuance from leaf to root
+func validateCertChain(certChain []*x509.Certificate) error {
```
Comment (patrickzheng200) [inline signature/jws/jws.go:222]: See above. We should already have this implementation in the x509 package.

File: signature/jws/jws.go [hunk 4]
```diff
@@ -25,24 +35,374 @@ func NewEnvelope() signature.Envelope {
...
+			} else if nextCert := certChain[i+1]; !isIssuedBy(cert, nextCert) {
+				return fmt.Errorf("certificate with subject %q is not issued by %q", cert.Subject, nextCert.Subject)
+			}
+		}
+	}
+
+	return nil
+}
+
+func isSelfSigned(cert *x509.Certificate) bool {
```
Comment (patrickzheng200) [inline signature/jws/jws.go:244]: Same here.

File: signature/jws/jws.go [hunk 5]
```diff
@@ -25,24 +35,374 @@ func NewEnvelope() signature.Envelope {
...
+	return isIssuedBy(cert, cert)
+}
+
+func isIssuedBy(subject *x509.Certificate, issuer *x509.Certificate) bool {
```
Comment (patrickzheng200) [inline signature/jws/jws.go:248]: Same here.

File: signature/errors.go (hunk 1 above)
Comment (JeyJeyGao) [reply signature/errors.go:75]: Added.

File: signature/jws/jws.go (hunk 2 above)
Comment (JeyJeyGao) [reply signature/jws/jws.go]: Thanks! Updated.

File: signature/jws/jws.go (hunk 4 above)
Comment (JeyJeyGao) [reply signature/jws/jws.go]: Done.

File: signature/jws/jws.go (hunk 5 above)
Comment (JeyJeyGao) [reply signature/jws/jws.go]: Done.

File: signature/jws/utils.go [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+			crit = append(crit, elm.Key)
+		}
+	}
+
+	alg, err := getJWSAlgo(sigAlg)
+	if err != nil {
+		return nil, err
+	}
+
+	jwsProtectedHdr := jwsProtectedHeader{
```
[the remark this answers is not in the document]
Comment (JeyJeyGao) [reply signature/jws/utils.go]: Updated.

File: signature/jws/utils.go [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+	}
+
+	return mergeMaps(m, extAttrs), nil
+}
+
+func getJWSAlgo(alg signature.Algorithm) (string, error) {
+	jwsAlg, ok := signatureAlgJWSAlgMap[alg]
+	if !ok {
+		return "", signature.SignatureAlgoNotSupportedError{
+			Alg: signatureAlgJWSAlgMap[alg]}
```
[the remark this answers is not in the document]
Comment (JeyJeyGao) [reply signature/jws/utils.go]: Oh! Changed it to be fmt.Sprintf("#%d", alg)

File: signature/jws/jws.go (hunk 3 above)
Comment (JeyJeyGao) [reply signature/jws/jws.go]: Updated.

**Review by binbin-li — APPROVED** [review]
lgtm

**Review by patrickzheng200 — APPROVED** [review]
LGTM

**Pull request merged**